Swap JSON parser

Matthew-Mosior opened the pull request:
This PR swaps out the JSON parser found in `Language.JSON` in `contrib` for @stefan-hoeck's `parser-json` (`idris2-parser`).

This also remove the dependency on `contrib`.

Matthew-Mosior committed:
Swapping out contrib JSON parser for parser-json (idris2-parser) JSON parser, and removing contrib dependency.

Matthew-Mosior committed:
Updating Makefile for addition of parser-json library.

Matthew-Mosior committed:
Fixing Makefile.

Matthew-Mosior committed:
Actually fixing Makefile (hopefully).

mattpolzin commented:
The Makefile ought to run locally for you to save the round trip of GitHub CI running, though I don’t mind hitting the button to get it to run again.

Matthew-Mosior commented:
> The Makefile ought to run locally for you to save the round trip of GitHub CI running, though I don’t mind hitting the button to get it to run again.

I apologize, was fairly certain this Makefile would build.

I did `make` locally first and it built.  Not sure what is going on here, but will get this fixed up.

mattpolzin commented:
Without digging into the changes in much depth yet, my high level critique would be that I assume there's no reason we can't retain totality for those functions you've made partial so I'd like to do that since totality is such a nice guarantee. Even when I don't get totality I almost never opt for partial over covering (again because coverage checking is such a nice benefit offered by the compiler).

Matthew-Mosior commented:
> Without digging into the changes in much depth yet, my high level critique would be that I assume there's no reason we can't retain totality for those functions you've made partial so I'd like to do that since totality is such a nice guarantee. Even when I don't get totality I almost never opt for partial over covering (again because coverage checking is such a nice benefit offered by the compiler).

I agree with that, I'll re-work the `partial` functions so that they are at least `covering`.

Matthew-Mosior committed:
Removing partial annotations by swapping out the strTail with substr.

Matthew-Mosior committed:
Removing extraneous whitespace.

Matthew-Mosior committed:
Removing extraneous whitespace.

Matthew-Mosior committed:
Removing extraneous whitespace.

Matthew-Mosior committed:
Fixing MakeFile again (fingers crossed).

Matthew-Mosior commented:
@mattpolzin 

I was able to rewrite the `typeSpec` function using `substr` in `Prelude.Types` instead of using the `partial` `strTail` function in `Data.String`, removing all of the `partial` annotations.

See 47e01888693e879da76afba3076001165fcbad97.

Matthew-Mosior committed:
Trying to fix Makefile.

Matthew-Mosior commented:
@mattpolzin 

Not sure if you have time, but if this most recent commit doesn't build, would you mind taking a look at the Makefile?  I'm sure I'm missing something silly (not a Makefile expert by any means).

mattpolzin commented:
Sure. These days with pack around having a Makefile is kind of extra credit but I do want to maintain that because I’ve gone through the effort of making the project build without pack as a dependency. Happy to fix any remaining finicky details.

Matthew-Mosior commented:
> Sure. These days with pack around having a Makefile is kind of extra credit but I do want to maintain that because I’ve gone through the effort of making the project build without pack as a dependency. Happy to fix any remaining finicky details.

Awesome, thank you for helping with this!

mattpolzin committed:
Makefile dependency fixes

mattpolzin committed:
test updates

mattpolzin committed:
try different method entirely

mattpolzin committed:
but also look in system install location

mattpolzin committed:
prepare reusable test helper

mattpolzin committed:
support the check use-case. see about run helper not finding input file

mattpolzin committed:
remove contrib from test lib

mattpolzin committed:
undo unnecessary change that would have required updates to all expectation files

mattpolzin commented on src/Postgres/DB.idr:
```diff
@@ -295,7 +295,6 @@ nextNotification = pgExec pgNextNotification
 ||| in your programs existing logic loop unless your entire loop
 ||| is dictated by notification arrival anyway.
 export
-partial
```
This should not be removed. This function does not necessarily make progress because it blocks on external notification arrival (and there's no guarantee the next notification will ever arrive).

mattpolzin replied:
That said, `covering` is probably more appropriate than `partial` in hindsight.

mattpolzin commented on src/Postgres/Data/PostgresTable.idr:
```diff
@@ -303,9 +302,9 @@ insert : {n : _}
-      columnNames     = '(' <+ (join "," $ show . .name <$> (toList cols)) +> ')'
-      values = values table cols vs
-      valueStrings    = '(' <+ (join "," $ values) +> ')'
+      columnNames     = (strCons '(' (joinBy "," $ show . .name <$> (toList cols))) ++ (singleton ')')
```
I wonder if `singleton` is worth its verbosity relative to just using `")"`.

mattpolzin commented on src/Postgres/Data/PostgresValue.idr:
```diff
@@ -132,7 +132,12 @@ PGCast PString String where
+      Left _      =>
+        Nothing
+      Right json' =>
+        Just json'
```
```suggestion
    eitherToMaybe $ parseJSON Virtual (rawString json)
```
I'd prefer this in the other places where a similar case statement was written as well.

mattpolzin commented on src/Postgres/LoadTypes.idr:
```diff
@@ -126,7 +125,7 @@ parseType type = case isElem True typeSearch of
     typeSpec = if "_" `isPrefixOf` type
-                  then (True, drop 1 type)
+                  then (True, substr 1 ((length type) `minus` 1) type)
                   else (False, type)
```
Fortunately even though we don't have `drop` anymore, there's a newer even more appropriate option in `base` given the overall goal here.
```idris
    typeSpec = case (strM type) of
                    (StrCons '_' type') => (True, type')
                    _ => (False, type)
```

mattpolzin reviewed (commented):
Thanks for taking the time to swap out `contrib` with Stefan's faster JSON library!

I'm still working through the combination of build & test failures as you can probably see but while I work on that there are some stylistic/preferential requests I have for you.

mattpolzin committed:
ignore so library produced on Linux machines

mattpolzin committed:
specify the remaining required package for checking the readme

mattpolzin committed:
update README

mattpolzin committed:
bump versions, update example to not use contrib

mattpolzin committed:
more README updates

mattpolzin committed:
fix example after dropping contrib

Matthew-Mosior committed:
Addressing comments.

Matthew-Mosior committed:
Merge branch 'swap-json-parser' of github.com:Matthew-Mosior/pg-idris into swap-json-parser

mattpolzin committed:
Add jsonQuery test

Matthew-Mosior committed:
Addressing comments.

mattpolzin commented on src/Postgres/LoadTypes.idr:
```diff
@@ -68,7 +67,7 @@ oidTypeStrings : List String
 
 quote : String -> String
-quote str = '\'' <+ str +> '\''
+quote str = (strCons '\'' str) ++ "\\"
```
This used to put single quotes around the string but you've changed the closing quote to a backslash.

Matthew-Mosior committed:
Addressing comments.

Matthew-Mosior committed:
Addressing comments.

mattpolzin reviewed (approved):
I think we're there! Thanks again.

Matthew-Mosior commented:
> Thanks for taking the time to swap out `contrib` with Stefan's faster JSON library!
> 
> I'm still working through the combination of build & test failures as you can probably see but while I work on that there are some stylistic/preferential requests I have for you.

You're welcome, I'm glad that I am able to contribute!

Sounds good, see my commits below addressing your comments (you had great ideas/requests btw).

Matthew-Mosior commented:
> I think we're there! Thanks again.

Awesome, you're welcome!
